transports/janus_pfunix.c

```diff
@@ -0,0 +1,532 @@
+}
+
+gboolean janus_pfunix_is_janus_api_enabled(void) {
+	return pfd > 0;
+}
+
+gboolean janus_pfunix_is_admin_api_enabled(void) {
+	return admin_pfd > 0;
+}
+
+int janus_pfunix_send_message(void *transport, void *request_id, gboolean admin, json_t *message) {
+	if(message == NULL)
+		return -1;
+	if(transport == NULL) {
+		g_free(message);
+		return -1;
+	}
+	/* Make sure this is related to a still valid Unix Sockets session */
+	janus_pfunix_client *client = (janus_pfunix_client *)transport;
+	janus_mutex_lock(&clients_mutex);
+	if(g_hash_table_lookup(clients, client) == NULL || client->fd <= 0) {
+		janus_mutex_unlock(&clients_mutex);
+		JANUS_LOG(LOG_WARN, "Outgoing message for invalid client %p\n", client);
+		g_free(message);
+		message = NULL;
+		return -1;
+	}
+	janus_mutex_unlock(&clients_mutex);
+	/* Convert to string and enqueue */
+	char *payload = json_dumps(message, JSON_INDENT(3) | JSON_PRESERVE_ORDER);
+	json_decref(message);
+	g_async_queue_push(client->messages, payload);
+	/* Notify the thread there's data to send */
+	char c;
+	write(write_fd[1], &c, 1);
+	return 0;
+}
+
+void janus_pfunix_session_created(void *transport, guint64 session_id) {
+	/* We don't care */
+}
+
+void janus_pfunix_session_over(void *transport, guint64 session_id, gboolean timeout) {
+	/* We only care if it's a timeout: if so, close the connection */
+	if(transport == NULL || !timeout)
+		return;
+	/* FIXME Should we really close the connection in case of a timeout? */
+	janus_pfunix_client *client = (janus_pfunix_client *)transport;
+	janus_mutex_lock(&clients_mutex);
+	if(g_hash_table_lookup(clients, client) != NULL && client->fd > 0) {
+		/* Shutdown the client socket */
+		shutdown(client->fd, SHUT_WR);
+	}
+	janus_mutex_unlock(&clients_mutex);
+}
+
+
+/* Thread */
+void *janus_pfunix_thread(void *data) {
+	JANUS_LOG(LOG_INFO, "Unix Sockets thread started\n");
+
+	int fds = 0;
+	struct pollfd poll_fds[1024];	/* FIXME Should we allow for more clients? */
+	char buffer[4096];
+
+	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+		/* Prepare poll list of file descriptors */
+		fds = 0;
+		/* Writeable monitor */
+		poll_fds[fds].fd = write_fd[0];
+		poll_fds[fds].events = POLLIN;
+		fds++;
+		if(pfd > 0) {
+			/* Janus API */
+			poll_fds[fds].fd = pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		if(admin_pfd > 0) {
+			/* Admin API */
+			poll_fds[fds].fd = admin_pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		/* Iterate on available clients, to see if we need to POLLIN or POLLOUT too */
+		janus_mutex_lock(&clients_mutex);
+		GHashTableIter iter;
+		gpointer value;
+		g_hash_table_iter_init(&iter, clients_by_fd);
+		while(g_hash_table_iter_next(&iter, NULL, &value)) {
+			janus_pfunix_client *client = value;
+			if(client->fd > 0) {
+				poll_fds[fds].fd = client->fd;
+				poll_fds[fds].events = POLLIN;
+				if(g_async_queue_length(client->messages) > 0) {
+					poll_fds[fds].events = POLLOUT;
```

Comment:
Ops, you're right: this is what I meant, but I'm actually replacing it. Will fix that.